StateTimeline: Show more precise duration

zoltanbedi opened the pull request:
**What this PR does / why we need it**:

The state timeline tooltip shows the duration to only 1 precision point which means that it can be confusing to see that the duration is 1 hour when it is 1 hour and 50 minutes. I've added 1 more precision point so it will show minutes as well. 

![Screenshot 2022-01-14 at 14 44 13](https://user-images.githubusercontent.com/13729989/149525638-4d0679ea-edcd-4e42-b74c-44d075445434.png)

**Which issue(s) this PR fixes**:

Fixes #42269 

**Special notes for your reviewer**:

Should we make this customisable?

**To reproduce**:
<details>
<summary>Dashboard JSON</summary>

```
{
  "annotations": {
    "list": [
      {
        "builtIn": 1,
        "datasource": "-- Grafana --",
        "enable": true,
        "hide": true,
        "iconColor": "rgba(0, 211, 255, 1)",
        "name": "Annotations & Alerts",
        "target": {
          "limit": 100,
          "matchAny": false,
          "tags": [],
          "type": "dashboard"
        },
        "type": "dashboard"
      }
    ]
  },
  "editable": true,
  "fiscalYearStartMonth": 0,
  "graphTooltip": 2,
  "id": 205,
  "links": [],
  "liveNow": false,
  "panels": [
    {
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "thresholds"
          },
          "custom": {
            "fillOpacity": 70,
            "lineWidth": 0
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "#EAB839",
                "value": 0
              }
            ]
          },
          "unit": "string"
        },
        "overrides": []
      },
      "gridPos": {
        "h": 9,
        "w": 12,
        "x": 0,
        "y": 0
      },
      "id": 2,
      "options": {
        "alignValue": "left",
        "legend": {
          "displayMode": "list",
          "placement": "bottom"
        },
        "mergeValues": false,
        "rowHeight": 0.9,
        "showValue": "always",
        "tooltip": {
          "mode": "single",
          "sort": "none"
        }
      },
      "targets": [
        {
          "csvContent": "time,value\n1642142562605, LOW\n1642147962605, HIGH\n1642154362605, LOW",
          "datasource": {
            "type": "testdata",
            "uid": "PD8C576611E62080A"
          },
          "refId": "A",
          "scenarioId": "csv_content"
        }
      ],
      "title": "Duration",
      "type": "state-timeline"
    }
  ],
  "refresh": "",
  "schemaVersion": 34,
  "style": "dark",
  "tags": [],
  "templating": {
    "list": []
  },
  "time": {
    "from": "now-12h",
    "to": "now"
  },
  "timepicker": {},
  "timezone": "",
  "title": "state timeline",
  "uid": "HD3XieDnk",
  "version": 7,
  "weekStart": ""
}
```
</details>

zoltanbedi committed:
StateTimeline: Show more precise duration

dprokop commented:
Don't have thoughts yet, but initial feedback is that this probably ain't enough. See the data below (states spanning over months), we are still missing the precision:

```csv
time,value
1609459200000, LOW
1617235200000, HIGH
1625491500000, LOW
1630501500000,HIGH
````

ryantxu commented:
> Maybe the precision should be controlled via field config?

Lets avoid if possible :)

> Or alternatively we could be slightly more clever and automate the precision based on the minmal step of the time field.

This reminds me of a LONG ago requested feature that never got merged :cry: 
https://github.com/grafana/grafana/pull/11878

![image](https://user-images.githubusercontent.com/705951/149556406-6195a46b-0b3c-4d2c-938e-a70934aef0fc.png)

leeoniya commented:
certainly feels like there should be some kind of equivalent to `.toPrecision(3)` for time durations.

we should try to do show something like 3 elements of `years, months, weeks, days, hours, minutes, secs, ms` based on the range. maybe Moment has something like this built in already, or https://github.com/hustcc/timeago.js, or we can just hand-roll a small implementation, similar to 

https://github.com/grafana/grafana/blob/7422789ec7ab665fd5d01e3b891f44ac427061c5/public/app/features/explore/Time.tsx#L14

or

https://github.com/grafana/grafana/blob/7422789ec7ab665fd5d01e3b891f44ac427061c5/packages/grafana-ui/src/components/uPlot/config/UPlotAxisBuilder.ts#L188

leeoniya commented:
here's my version (with years, months, weeks support):

```js
// adapted from https://gist.github.com/remino/1563878
function fmtDuration(milli) {
  if (milli < 0 || Number.isNaN(milli)) {
    return null;
  }

  let yr, mo, wk, d, h, m, s, ms;

  s = Math.floor(milli / 1000);
  m = Math.floor(s / 60);
  s = s % 60;
  h = Math.floor(m / 60);
  m = m % 60;
  d = Math.floor(h / 24);
  h = h % 24;

  yr = Math.floor(d / 365);
  if (yr > 0)
    d = d % 365;

  mo = Math.floor(d / 30.5);
  if (mo > 0)
    d = d % 30;

  wk = Math.floor(d / 7);
  if (wk > 0)
    d = d % 7;

  ms = Math.floor((milli % 1000) * 1000) / 1000;

  return (
    yr > 0 ? yr + 'y '  + (mo > 0 ? mo + 'mo ' : '') + (wk > 0 ? wk + 'w ' : '') + (d > 0 ? d  + 'd ' : '') :
    mo > 0 ? mo + 'mo ' + (wk > 0 ? wk + 'w '  : '') + (d  > 0 ? d  + 'd ' : '') :
    wk > 0 ? wk + 'w '  + (d  > 0 ? d  + 'd '  : '') :
    d  > 0 ? d  + 'd '  + (h  > 0 ? h  + 'h '  : '') :
    h  > 0 ? h  + 'h '  + (m  > 0 ? m  + 'm '  : '') :
    m  > 0 ? m  + 'm '  + (s  > 0 ? s  + 's '  : '') :
    s  > 0 ? s  + 's '  + (ms > 0 ? ms + 'ms ' : '') :
    ms > 0 ? ms + 'ms ' :
    '0'
  ).trim();
}
```

zoltanbedi committed:
Add Leon's calculation

zoltanbedi commented:
@leeoniya I added your code to the state timeline's util file and changed it slightly for months to calculate with 28 days instead of 30.5. Let me know if you think that doesn't make sense.

leeoniya commented on public/app/plugins/panel/state-timeline/utils.ts:
```diff
@@ -567,3 +567,66 @@ export function findNextStateIndex(field: Field, datapointIdx: number) {
+  }
+
+  ms = Math.round((milliSeconds % 1000) * 1000) / 1000;
+  console.log({ ms, s, m, h, d, wk, mo, yr });
```
oops

leeoniya commented on public/app/plugins/panel/state-timeline/utils.ts:
```diff
@@ -567,3 +567,66 @@ export function findNextStateIndex(field: Field, datapointIdx: number) {
+    d = d % 365;
+  }
+
+  console.log({ d });
```
oops

zoltanbedi replied:
oops 😄

leeoniya commented:
> calculate with 28 days instead of 30.5. Let me know if you think that doesn't make sense.

11 * 28 is quite short of 11 months. i think 30 or 30.5 makes more sense here. 🤷 

also might be worth adding a couple tests that skip a segment. e.g. "1y 1w" instead of what we had before like "1y 0mo 1w". this code was made to omit any 0 intermediates.

do you think this util fn should be moved out to some general date utils or keep local to this vis?

zoltanbedi committed:
Change to 30 days

Remove console logs. Add more tests.

zoltanbedi commented:
> > calculate with 28 days instead of 30.5. Let me know if you think that doesn't make sense.
> 
> 11 * 28 is quite short of 11 months. i think 30 or 30.5 makes more sense here. 🤷
> 
> also might be worth adding a couple tests that skip a segment. e.g. "1y 1w" instead of what we had before like "1y 0mo 1w". this code was made to omit any 0 intermediates.
> 
> do you think this util fn should be moved out to some general date utils or keep local to this vis?

I choose 28 days as it was kind of weird for me that for 28 days it says that 4 weeks instead of 1 month but 30 days makes sense for me as well.

Added some more tests.

So I was thinking about this and the reason I choose to put it next to the vis is because we already have a toDuration function with parameters. People would be confused which one to use. But I'm up for moving this out from here.

leeoniya commented:
> So I was thinking about this and the reason I choose to put it next to the vis is because we already have a toDuration function with parameters. People would be confused which one to use. But I'm up for moving this out from here.

sounds good. toDuration() could use a refactor and maybe we can merge these in the future.

dprokop reviewed (approved):
Now it works way better @zoltanbedi!
